src/storage/src/source/util.rs

```diff
@@ -144,3 +146,67 @@ where
     // so we are guaranteed that `token` is non-None.
     (data_stream, token.unwrap())
 }
+
+/// lalal
+pub fn async_source<G, D, B, L>(
+    scope: &G,
+    name: String,
+    input: Stream<G, ()>,
+    construct: B,
+) -> (Stream<G, D>, AsyncSourceToken)
+where
+    G: Scope<Timestamp = Timestamp>,
+    D: Data,
+    B: FnOnce(
+        OperatorInfo,
+        CapabilitySet<Timestamp>,
+        AsyncInputHandle<
+            Timestamp,
+            Vec<()>,
+            <Pipeline as timely::dataflow::channels::pact::ParallelizationContractCore<
+                G::Timestamp,
+                Vec<()>,
+            >>::Puller,
+        >,
+        OutputWrapper<G::Timestamp, Vec<D>, TeeCore<G::Timestamp, Vec<D>>>,
+    ) -> L,
+    L: Future + 'static,
+{
+    let mut token = None;
+
+    let mut builder = AsyncOperatorBuilder::new(name, scope.clone());
+    let operator_info = builder.operator_info();
+
+    let (data_output, data_stream) = builder.new_output();
+
+    let remap_input = builder.new_input_connection(
+        &input,
+        Pipeline,
+        // As documented, the optional input does not
+        // participate in progress tracking.
+        vec![Antichain::new()],
+    );
+
+    builder.build(|capabilities| {
+        let cap_set = CapabilitySet::from_elem(capabilities.into_element());
+
+        let (tx, mut rx) = tokio::sync::oneshot::channel();
```

Comment:
can we construct this outside the build closure and only move the sender into it?